core/models.py

```diff
@@ -0,0 +1,125 @@
+import hashlib
+import re
+from typing import ClassVar
+
+from django.contrib.auth import get_user_model
+from django.contrib.contenttypes.fields import GenericForeignKey, GenericRelation
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+User = get_user_model()
+
+
+class GenericModel(models.Model):
+    user = models.ForeignKey(User, on_delete=models.SET_NULL, null=True)
+    is_active = models.BooleanField(default=True)
+    is_locked = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:  # noqa: D106
+        abstract = True
+
+
+class Community(GenericModel):
+    name = models.CharField(max_length=255)
+
+    def __str__(self) -> str:  # noqa: ANN101, D105
+        return str(self.name)
+
+
+class Tag(models.Model):
+    name = models.SlugField()
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey("content_type", "object_id")
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    class Meta:  # noqa: D106
+        indexes: ClassVar[list[tuple[str, ...]]] = [
+            models.Index(fields=["content_type", "object_id"]),
+        ]
+
+    def __str__(self) -> str:  # noqa: ANN101, D105
+        return str(self.name)
+
+
+class Post(GenericModel):
+    community = models.ForeignKey(
+        Community,
+        on_delete=models.CASCADE,
+        related_name="posts",
+    )
+    title = models.CharField(max_length=255)
+    content = models.TextField()
+
+    tags = GenericRelation(Tag, related_query_name="posts")
+    image = models.ImageField(blank=True)
+    parent = GenericRelation("self", related_query_name="children")
+    version: str = models.CharField(max_length=32)
+
+    def __str__(self) -> str:  # noqa: ANN101, D105
+        return f"@{self.user}: {self.title}"
+
+    def save(self, *args, **kwargs) -> None:  # noqa: ANN101, ANN002, ANN003
+        super().save(*args, **kwargs)
+
+        current_tags: set[str] = set(re.findall(r"#(\w+)", self.content))
+        existing_tags: set[str] = set(
+            Tag.objects.filter(
+                content_type=self.get_content_type(),
+                object_id=self.id,
+            ).values_list("name", flat=True),
+        )
+
+        # Remove tags that were removed in the content
+        tags_to_remove = existing_tags - current_tags
+        Tag.objects.filter(
+            name__in=tags_to_remove,
+            content_type=self.get_content_type(),
+            object_id=self.id,
+        ).delete()
+
+        # Add new tags
+        new_tags = current_tags - existing_tags
+        for tag in new_tags:
+            Tag.objects.create(name=tag, content_object=self)
+
+        if self.created_at != self.updated_at:
+            self.version = self.generate_version()

```

Comment:
This is not saved. 
Moreover, it will create new version on every save click after first update, even if there is no change. 
Actually, I am thinking this condition is not needed, no? Worst case it will generate md5 again, but since no change in text, we loose bit compute power on hash generation. 
Either way, it needs save, so prob move before super.save()